{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "160ac9fb-2380-485a-924b-aff1c29f51e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install pycirclize pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5daa945-f1c1-4170-b3a7-0322436514f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from pycirclize import Circos\n",
    "from pycirclize.parser import Genbank\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dbf2cef8-86ff-4cc2-a9c5-b43692fe210f",
   "metadata": {},
   "outputs": [],
   "source": [
    "class MyRadius:\n",
    "    def __init__(self):\n",
    "        self._radius = 100\n",
    "\n",
    "    def get(self, thin=False):\n",
    "        width = 4 if thin else 8\n",
    "        value = self._radius - width, self._radius\n",
    "        self._radius -= width\n",
    "        return value\n",
    "\n",
    "def get_name(old):\n",
    "    return '-'.join(old.split('-')[1:])\n",
    "\n",
    "\n",
    "def get_name2(old):\n",
    "    return old.split('-')[0]\n",
    "\n",
    "\n",
    "def percent_to_float(old: str) -> float:\n",
    "    return float(old.rstrip('%')) / 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d55c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read data\n",
    "# evaluation result table, remove abnormal genes\n",
    "count_threshold = 10\n",
    "filename = r'rodents_Evaluation.csv'\n",
    "gb_file = r'extend.gb'\n",
    "gb = Genbank(gb_file)\n",
    "data_raw = pd.read_csv(filename)\n",
    "data_raw['Name'] = data_raw.Loci.apply(get_name)\n",
    "data_raw['Loci'] = data_raw.Loci.apply(get_name2)\n",
    "data_raw['Gap_Ratio'] = data_raw['Gap_Ratio'].apply(percent_to_float)\n",
    "data_raw['Tree_Res'] = data_raw['Tree_Res'].apply(percent_to_float)\n",
    "data_raw['Total_GC'] = data_raw['Total_GC'].apply(percent_to_float)\n",
    "data_raw['Observed_Res'] = data_raw['Observed_Res'].apply(percent_to_float)\n",
    "data_raw2 = data_raw[data_raw.Samples >= count_threshold]\n",
    "data = data_raw2[data_raw2.Loci != 'gene']\n",
    "data_names_raw = data.Name.tolist()\n",
    "data_names = list()\n",
    "for i in data_names_raw:\n",
    "    new = re.sub('trn(.|fM)_...', r'trn\\1', i)\n",
    "    data_names.append(new)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff3c0f93",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reference organelle genome genbank file from OGU.gb2fasta\n",
    "# total length - ir length\n",
    "r = MyRadius()\n",
    "# yellow,yellow/gray,blue,blue,green?\n",
    "feature_colors = '#FFD93D,#DDDDDD,#4D96FF,#9D96FF,#6BCB77'.split(',')\n",
    "track_colors = list(\n",
    "    reversed('#F7931B,#66B3FF,#338CFF,#3F51B5,#FB9883,#CC5500,#FF6347,#36454F,#778877'.split(',')))\n",
    "features = ('CDS', 'intron', 'tRNA', 'rRNA', 'spacer')\n",
    "colname_text = (('Tree_Res', 'Tree resolution'), ('PD', 'PD'), ('PD_stem', 'PD-stem'),\n",
    "                ('PD_terminal', 'PD-terminal'),\n",
    "                ('Observed_Res', 'Observed resolution'), ('Pi', 'Pi'), ('Entropy', 'Shannon index'),\n",
    "                ('Total_GC', 'GC ratio'),\n",
    "                ('Gap_Ratio', 'Gap ratio'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b007c9-bdb5-41f6-abfa-8ebd90b138e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(12, 12))\n",
    "# circos = Circos(sectors, space=0.5)\n",
    "# genome gb file as template, no extra treatmetn\n",
    "circos = Circos(sectors={gb.name: gb.range_size},\n",
    "                sector2clockwise={gb.name: False})\n",
    "sector = circos.sectors[0]\n",
    "r1 = r.get()\n",
    "feature_track = sector.add_track(r1)\n",
    "feature_track.axis(fc=\"#FFFFFF\", ec=\"none\")\n",
    "gene_labels = []\n",
    "gene_list = []\n",
    "last_gene = ''\n",
    "gene_set = set()\n",
    "ir_pos_list = set()\n",
    "for feat in gb.extract_features(\"gene\"):\n",
    "    start, end = int(str(feat.location.start)), int(str(feat.location.end))\n",
    "    pos = (start + end) / 2\n",
    "    label = feat.qualifiers.get(\"gene\", [\"??\"])[0]\n",
    "    if label == \"\" or label.startswith(\"hypothetical\") or label == last_gene:\n",
    "        continue\n",
    "    if len(label) > 20:\n",
    "        label = label[:20] + \"...\"\n",
    "    last_gene = label\n",
    "    gene_list.append(pos)\n",
    "    gene_labels.append(label)\n",
    "    if label in gene_set:\n",
    "        ir_pos_list.add(pos)\n",
    "    gene_set.add(label)\n",
    "\n",
    "# Plot CDS product labels on outer position\n",
    "feature_track.xticks(\n",
    "    gene_list,\n",
    "    gene_labels,\n",
    "    label_orientation=\"vertical\",\n",
    "    show_bottom_line=True,\n",
    "    label_size=6,\n",
    "    #line_kws=dict(ec=\"white\"),\n",
    ")\n",
    "# extract names\n",
    "pos_list, labels = [], []\n",
    "last_name = ''\n",
    "for feature, color in zip(features, feature_colors):\n",
    "    feature_track.genomic_features(\n",
    "        gb.extract_features(feature),\n",
    "        plotstyle=\"box\",\n",
    "        r_lim=r1,\n",
    "        fc=color, alpha=1, lw=0.1, ec='black')\n",
    "    for f in gb.extract_features(feature):\n",
    "        start, end = int(str(f.location.start)), int(str(f.location.end))\n",
    "        pos = end\n",
    "        if feature == 'intron':\n",
    "            label = f.qualifiers['gene'][0] + '.' + f.qualifiers['number'][0]\n",
    "        elif feature != 'spacer':\n",
    "            label = f.qualifiers.get(\"gene\", [\"??\"])[0]\n",
    "            if label == last_name:\n",
    "                continue\n",
    "        else:\n",
    "            label = f.qualifiers['upstream'][0] + '-' + f.qualifiers['downstream'][0]\n",
    "        last_name = label\n",
    "        pos_list.append(pos)\n",
    "        labels.append(label)\n",
    "feature_track.genomic_features(gb.extract_features('D-loop'), plotstyle='box', r_lim=r1,\n",
    "                              fc='#999999', alpha=1, lw=0.2, ec='black')\n",
    "labels_set = set(labels)\n",
    "data_names_set = set(data_names)\n",
    "#a_b = sorted(list(data_names_set-labels_set))\n",
    "#print('data_names_set-labels_set', a_b, len(a_b))\n",
    "#print('\\n'*3)\n",
    "#b_a = sorted(list(labels_set-data_names_set))\n",
    "#print('labels_set-data_names_set', b_a, len(b_a))\n",
    "intersection = data_names_set & labels_set\n",
    "clean_data = data.query('Name in @intersection')\n",
    "label_pos = list()\n",
    "label_pos_dict = dict()\n",
    "ir_second_label_pos = list()\n",
    "for key, value in zip(labels, pos_list):\n",
    "    if key not in intersection:\n",
    "        continue\n",
    "    label_pos.append((key, value))\n",
    "    label_pos_dict[key] = value\n",
    "\n",
    "clean_data['Position'] = [label_pos_dict[x] for x in clean_data.Name]\n",
    "clean_data = clean_data.sort_values(by=['Position'])\n",
    "clean_pos = clean_data.Position.tolist()\n",
    "widths = list()\n",
    "widths.append(clean_pos[0])\n",
    "for i in range(1, len(clean_pos)):\n",
    "    w = clean_pos[i] - clean_pos[i - 1]\n",
    "    widths.append(w)\n",
    "\n",
    "#print(clean_data[['Name','Position','Loci']].head, clean_data.shape, len(clean_pos))\n",
    "#print(clean_data.Name.tolist(), clean_pos)\n",
    "#print(ir_second_label_pos)\n",
    "def draw_bar(track, x_, y_, w_, text):\n",
    "    max_y = max(y_)\n",
    "    color = track_colors.pop()\n",
    "    for x, y, w in zip(x_, y_, w_):\n",
    "        # bar need x[list], y[list]\n",
    "        # x should adjusted by width\n",
    "        track.bar([x - w / 2], [y], width=w, color=color, \n",
    "                  vmin=0, vmax=max_y)\n",
    "\n",
    "    # ec='black', lw=0.5\n",
    "\n",
    "\n",
    "for col, text in colname_text:\n",
    "    track = sector.add_track(r.get(), r_pad_ratio=0.1)\n",
    "    track.axis()\n",
    "    col_data = clean_data[col].tolist()\n",
    "    draw_bar(track, clean_pos, col_data, widths, text)\n",
    "\n",
    "# print(clean_data.columns)\n",
    "\n",
    "fig = circos.plotfig()\n",
    "\n",
    "track_colors2 = '#F7931B,#66B3FF,#338CFF,#3F51B5,#FB9883,#CC5500,#FF6347,#36454F,#778877'.split(',')\n",
    "rect_handles = []\n",
    "for name, color in zip(features, feature_colors):\n",
    "    rect_handles.append(Patch(color=color, label=name))\n",
    "rect_handles.append(Patch(color='#999999', label='D-loop'))\n",
    "for name, color in zip(colname_text, track_colors2):\n",
    "    rect_handles.append(Patch(color=color, label=name[1]))\n",
    "_ = circos.ax.legend(handles=rect_handles, bbox_to_anchor=(1, 0),\n",
    "                     loc='best', fontsize=8, title=\"Types\", ncol=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dedd4df8-c69c-4faf-b46a-301817ad2eb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.savefig('out.pdf')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
